### PR TITLE
added httpx opentelemetry adapter

### PR DESCRIPTION
**Pull request by HoneyryderChuck**
[httpx](https://gitlab.com/os85/httpx) is a ruby http client.

### CODE REVIEW

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/instrumentation.rb
Comment (arielvalentin) [inline instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/instrumentation.rb:27]: Would you be able to help me understand why this code is replacing the Session class?

**Review by arielvalentin — COMMENTED** [review]
Thank you again for your contribution! We really appreciate your desire to add instrumentation to HTTPX. 

_This is the first time I am learning about the gem and I think it has a lot of essential features for modern Ruby applications. I most definitely would like to give it a try!_
 
Since you are the inventor and maintainer of HTTPX, would you have any interest in being one of the first gem authors to provide First Party support and make OTel one of the official plugins available for your gem?

We could work with you to get the gem listed in the official registry :)

**Comment by HoneyryderChuck** [comment]
Thx!

> would you have any interest in being one of the first gem authors to provide First Party support and make OTel one of the official plugins available for your gem?

What does that entail? FWIW httpx already ships with adapters to other tools, including [one for datadog](https://honeyryderchuck.gitlab.io/httpx/wiki/Datadog-Adapter). It seems that in OTel's case every integration ships as its own gem, so not sure if worth deviating frrom that path?

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/instrumentation.rb
Comment (HoneyryderChuck) [reply instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/instrumentation.rb:27]: httpx ships with its own plugin system, where one can load functionality on an instance without affecting the others: `session_with_retries = HTTPX.plugin(:retries)`. However, in the case of OTel and other adapters, the behaviour needs to be loaded by default for all instances, so the default session class is overridden by default + otel plugin. This looks ugly, but it's a require-time pass.

Comment (arielvalentin) [reply instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/instrumentation.rb:27]: Thanks for the explanation. As long as you are comfortable with this approach I understand why this needs to be done this way. 

Do other tracing plugins do something similar? E.g. DataDog Adapter you shared in one of your other comments?

**Comment by arielvalentin** [comment]
> Thx!
> 
> > would you have any interest in being one of the first gem authors to provide First Party support and make OTel one of the official plugins available for your gem?
> 
> What does that entail? FWIW httpx already ships with adapters to other tools, including [one for datadog](https://honeyryderchuck.gitlab.io/httpx/wiki/Datadog-Adapter). It seems that in OTel's case every integration ships as its own gem, so not sure if worth deviating frrom that path?

Longer term I think our goal is to have First Party Integration with other gems. We have hesitated doing this in the past because the instrumentation-wg is still immature and semantic conventions is still experimental in quite a few cases. 

I am keen on giving first party support soon so that we are able to encourage gem authors to provide first-class support for OTel. We have limited bandwidth and lack a critical mass of experts to have a centralized repo for instrumentations.

This will require maintaining a close partnership with you and other httpx maintainers to ensure we are not inadvertently introducing unreasonable changes.   

As for the mechanics of making this happen, here is what I do to create my own instrumentations for our internal libraries at GitHub:

1. Move the instrumentation into the `httpx` repository. Your instrumentation should inherit from `OpenTelemetry::Instrumentation::Base`, which should already be done
2. Namespace the gem in your repository under the `httpx` namespace (as opposed to using `OpenTelemetry::Instrumentation`) 
3. Provide instructions to include the gem in the application's Gemfile. 
4. Bundler should take it from there!

The Base gem should register your instrumentation in the registry so there is no need to manually active it if the end user is relying on `use_all`. 

## All that being said

I am by no means requiring you to do any of this! 

AFAIK, this is the first time a gem author/maintainer is contributing an instrumentation to this repository so I was hoping you may be willing to give a try to First Party instrumentation.

We welcome this contribution and will add you as a CODEOWNER so that we are able to contact you for on-going maintenance in this repo.

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/plugin.rb
Comment (arielvalentin) [inline instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/plugin.rb:62]: Is there a case where you will not have a trace token but _will_ have a span?

Does httpx's plugin system guard against errors raised in individual plugins? 

The `detach` method raises errors in cases where the `token` is popped out of order.

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/version.rb
Comment (arielvalentin) [inline instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/version.rb:10]: Instrumentations have their own versioning scheme. The release system will generate an initial release version of `0.1.0` and auto-increment with every release based on the use of conventional commits. 

```suggestion
      VERSION = '0.1.0'
```

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/instrumentation.rb
Comment (arielvalentin) [inline instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/instrumentation.rb:11]: Based on the Appraisals, it looks like you are only testing versions `0.24` and newer.

I recommend adding a runtime check here for a min gem version of `0.24` in a custom `compatible` block so that users who are on older versions of the gem know they have to upgrade in order to use this instrumentation.

**Review by arielvalentin — CHANGES_REQUESTED** [review]

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/instrumentation.rb
Comment (HoneyryderChuck) [reply instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/instrumentation.rb:31]: yes, all the other plugins go about a similar approach 👍

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/plugin.rb
Comment (HoneyryderChuck) [reply instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/plugin.rb:32]: I understand, but I see no immediate benefit from deviating from the standard (didn't see any other adapter not using it). I'd say it's for the OTel team to remove usage of the gem and update code accordingly. Hope you agree.

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/plugin.rb
Comment (HoneyryderChuck) [reply instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/plugin.rb]: done

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/plugin.rb
Comment (HoneyryderChuck) [reply instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/plugin.rb:52]: yes, HTTP status (even 4xx and 5xx) will be wrapped in `HTTPX::Response`.

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/plugin.rb
Comment (HoneyryderChuck) [reply instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/plugin.rb:61]: this is just a guard against the possibility that an error is raised in the function above before the @trace_token is initialized.

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/version.rb
Comment (HoneyryderChuck) [reply instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/version.rb]: done

File: instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/instrumentation.rb
Comment (HoneyryderChuck) [reply instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/instrumentation.rb:11]: made the change.

**Comment by HoneyryderChuck** [comment]
I've no problem with providing support for this adapter. I'm not sure what you mean as "first party" though. Does it mean moving the adapter into its own repo? I'm not sure if that would be beneficial at the stage OTel ruby is. But feel free to ping me in any issue and question related to it.

**Comment by arielvalentin** [comment]
> I'm not sure what you mean as "first party" though. Does it mean moving the adapter into its own repo?

Ideally OTel support would be a "first party" instrumentation, meaning maintained by gem authors; and not "third party" meaning maintained in this repository by us. 

In this case it would mean residing with the other adapters and included in the httpx like the others are: https://gitlab.com/os85/httpx/-/tree/master/lib/httpx/adapters?ref_type=heads

Then in that ideal world, we would change the instrumentation portion of this repo to provide utilities and tools for gem maintainers. 

The challenge we face with third party maintenance is that we have limited capacity and expertise and when upstream changes happen with gems being instrumented in this repo. We have had a few contributions and instrumentations that do not have active maintainers and at some point in the future we will not be able to keep up and have to sunset them.

Hope that clears up what I mean.

**Comment by arielvalentin** [comment]
@HoneyryderChuck There are a few more steps to ensure tests are passing on multiple versions of Ruby: 

Please add the instrumentation to the CI builds: 
- https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/.github/workflows/ci-instrumentation.yml#L16
- https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/.github/workflows/ci-instrumentation-canary.yml#L13

Please add the gem to the release list:
- https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/.toys/.data/releases.yml#L193

Once all of the tests are passing I will merge.

File: instrumentation/httpx/CHANGELOG.md
Comment (arielvalentin) [inline instrumentation/httpx/CHANGELOG.md:3]:
```suggestion
### v0.1.0 / 2023-10-10
```

File: instrumentation/httpx/CHANGELOG.md
Comment (arielvalentin) [inline instrumentation/httpx/CHANGELOG.md:2]: The Release automation will generate the Changelog for you. It is very strict and will not allow manual edits to this file. 

```suggestion
```

**Comment by HoneyryderChuck** [comment]
@arielvalentin updated

**Comment by arielvalentin** [comment]
Looks like there are failing specs: https://github.com/open-telemetry/opentelemetry-ruby-contrib/actions/runs/6626632507/job/18000606095?pr=681

**Comment by HoneyryderChuck** [comment]
fixed.

File: instrumentation/httpx/CHANGELOG.md
Comment (HoneyryderChuck) [reply instrumentation/httpx/CHANGELOG.md]: done

**Comment by HoneyryderChuck** [comment]
@arielvalentin any updates on this one? do you think it's ready?

**Review by arielvalentin — APPROVED** [review]

File: instrumentation/httpx/README.md
Comment (simi) [inline instrumentation/httpx/README.md:35]: Should this be `opentelemetry-instrumentation-httpx`?

Comment (arielvalentin) [reply instrumentation/httpx/README.md:35]: Good catch @simi

Comment (arielvalentin) [reply instrumentation/httpx/README.md:35]:
```suggestion
The `opentelemetry-instrumentation-httpx` gem source is [on github][repo-github], along with related gems including `opentelemetry-api` and `opentelemetry-sdk`.
```

File: instrumentation/httpx/README.md
Comment (arielvalentin) [inline instrumentation/httpx/README.md:41]:
```suggestion
The `opentelemetry-instrumentation-httpx` gem is distributed under the Apache 2.0 license. See [LICENSE][license-github] for more information.
```

**Review by simi — APPROVED** [review]

**Pull request merged**
